### ISSUE TITLE
Raspberry Pi (ARM-Linux) H2R Graphics Release

### ISSUE THREAD

**Issue by MarcusAlzona** [iamjohnbarker/h2r-graphics-releases#13]
In terms of hardware form factors, the Raspberry Pi may be one of the most ideal platforms to (potentially) run H2R Graphics.  Would it be possible to build H2R for the Raspberry Pi?  (assuming the current Linux AppImage isn't ARM)

The capability to run "headless" would be ideal - having the Pi auto-run H2R, displaying the Chroma Window fullscreen (without a window manager) and having control done by OSC + ssh.  I would suggest a primary target would be for the Raspberry Pi 4, given its dual-HDMI output capability and more robust CPU/memory specifications.  

Please let me know if there's anything that could be done to help with this effort.  Thank you!

**Comment by iamjohnbarker** [comment]
Working on this indeed and keen to get it going in a future major release!

**Comment by mkotek** [comment]
If there was a way to actually combine it with a Companion Pi image, that would be even better - it would require to have just 1 RPi next to ATEM.

**Comment by bevand10** [comment]
@iamjohnbarker @mkotek 

This could be worked on by the community if this project was not closed-source.

**Comment by iamjohnbarker** [comment]
The major reworking of the application is my current focus so I am working towards great feature like this.

**Comment by aeciolevy** [comment]
That would be amazing having H2R Graphics + PlayoutBee + Companion in an RPI 4...

**Comment by mathias-ewald** [comment]
@iamjohnbarker Have you made any progress here?

**Comment by iamjohnbarker** [comment]
Noting new, sadly. Turns out to be quite the effort and I am still focused on making 2.0 as best it can be. Thanks for the reminder though!

**Comment by MarcusAlzona** [comment]
Hello!

I was looking at re-doing a few items in the streaming setups I support, and remembered the (possibility) of incorporating H2R into the setup.  While I could bring an additional laptop just to run H2R, having H2R running on a Pi...in particular a Pi that is also running Companion - would be an incredibly powerful embedded package.

Think RetroPie, H2R style....

Given that you're about a year out from your 2.0 release, any thoughts on compiling this for a Pi4 @iamjohnbarker ?  Is there anything we can do to help?  Maybe if you get an alpha or beta compiled, we can help try to figure out how to get it to auto-run on a headless screen, so if you plug it in it already is sending a 1080p signal or something...

Just thought I'd check in on this two year old issue (lucky number 13) 😄 

Thanks!

**Comment by iamjohnbarker** [comment]
@MarcusAlzona Thanks for the follow up!

I have looked into this a few times as I wanted to make a downloadable image that folks could run and it would "Just Work".

Then I switched from that to just build a Pi installable app but it failed on build repeatedly and I just couldn't keep spending time on it.

Every once in a while I dig back in and try again though!

**Comment by bevand10** [comment]
Any reason you cannot open source the project and let the community contribute a fix to this issue?

Seems a bit odd that everything here relies on a single person.

**Comment by iamjohnbarker** [comment]
@bevand10 Good question and I do indeed have a reason - it's my full time job and full time income making these apps now so I am keen to keep growing them myself!

**Comment by MarcusAlzona** [comment]
> @MarcusAlzona Thanks for the follow up!
> 
> I have looked into this a few times as I wanted to make a downloadable image that folks could run and it would "Just Work".
> 
> Then I switched from that to just build a Pi installable app but it failed on build repeatedly and I just couldn't keep spending time on it.
> 
> Every once in a while I dig back in and try again though!

Completely understandable!  I believe we all (should) want you to keep on developing your business, as H2R is such a contribution to the community.

If you make any progress with the build (of just the app), please let me know.  I've been researching how to make the Pi boot into the graphical environment (X or otherwise) while HDMI is disconnected, in order to have it auto-launch into H2R and have the ability to connect/disconnect the physical device(Pi) as needed, and not get stuck on boot....

**Comment by iamjohnbarker** [comment]
@MarcusAlzona I'm happy to say I have an update - which is in experimental form.

The latest 2.15.1 beta build now has an armv7l version: https://github.com/iamjohnbarker/h2r-graphics-releases/releases/tag/v2.15.1

This can be executed on the Pi with the latest 32bit PiOS and should run nicely.

That said if you open an output on the Pi it isn't great, but over the network (like you have mentioned) you can go to `PI_IP_ADDRESS:4001` and use the app as expected!

**Comment by ovinas** [comment]
While 2.15.2 (armv7l) starts fine on my Pi3, 2.15.3 doesn't. Noticed a significant difference in file size. Maybe that's a hint.

**Comment by MarcusAlzona** [comment]
> @MarcusAlzona I'm happy to say I have an update - which is in experimental form.
> 
> The latest 2.15.1 beta build now has an armv7l version: https://github.com/iamjohnbarker/h2r-graphics-releases/releases/tag/v2.15.1
> 
> This can be executed on the Pi with the latest 32bit PiOS and should run nicely.
> 
> That said if you open an output on the Pi it isn't great, but over the network (like you have mentioned) you can go to `PI_IP_ADDRESS:4001` and use the app as expected!

Thank you!  Moved some Pi hardware around, and managed to test it over the last couple of days with multiple OS images and hardware combinations.

In the end, what worked was Raspberry Pi OS bullseye 32-bit (armhf), on both the Pi3 and Pi4/400.  Tried the 64-bit version (just in case), but it didn't run (as expected).

I tried to run the H2R armv7l AppImage on the Companion Pi v2.4.1+70 (Stable) and 0.0.0+111 (Beta), including following the instructions to install the Desktop User Interface ( `sudo apt install lightdm xfce4 xfce4-terminal rc-gui chromium filezilla` ).  Unfortunately....neither worked, as they are both 64-bit (aarch64).

(splitting into multiple comments to separate the discussion points)

**Comment by MarcusAlzona** [comment]
> While 2.15.2 (armv7l) starts fine on my Pi3, 2.15.3 doesn't. Noticed a significant difference in file size. Maybe that's a hint.

Yes - noticed this too.  Is this something specific to the code changes, or perhaps is this a build script / library requirement issue?

Would it be possible to get a 64-bit Pi version included in the build scripts?  Companion Pi requires a Pi4/400 4GB/8GB, and is unsupported on other variants.  I'm assuming there's an implicit 64-bit requirement as well (but I could be mistaken).

**Comment by MarcusAlzona** [comment]
> In the end, what worked was Raspberry Pi OS bullseye 32-bit (armhf), on both the Pi3 and Pi4/400. Tried the 64-bit version (just in case), but it didn't run (as expected).

In related news, I was able to get H2R Graphics to auto-start on boot!  There are multiple ways to do it, but the cleanest I've been able to figure out so far was to add an entry in the window manager auto start.  You copy/rename the AppImage to some "standard" name (that you'll update over time), then create a autostart directory and then an entry in that directory for the binary.

```
cd ~
wget https://github.com/iamjohnbarker/h2r-graphics-releases/releases/download/v2.15.2/H2R-Graphics-2.15.2-armv7l.AppImage
mv H2R-Graphics-2.15.2-armv7l.AppImage H2R.AppImage
mkdir .config/autostart
nano .config/autostart/H2R.desktop
```

...and in that H2R.desktop file:
```
[Desktop Entry]
Type=Application
Name=H2R
Exec=/home/pi/H2R.AppImage
``` 
Let me know if this works for you all!

**Comment by ovinas** [comment]
2.15.3 (armv7l) is fixed and now starts fine.

**Comment by ferramarco** [comment]
Hi, I have an OrangePi 5, I already installed companion, is it possible to install H2R?

**Comment by iamjohnbarker** [comment]
> Hi, I have an OrangePi 5, I already installed companion, is it possible to install H2R?

I've not tried the OrangePi myself, however if you are able to install AppImages then you should be good @ferramarco!

**Comment by MarcusAlzona** [comment]
> > While 2.15.2 (armv7l) starts fine on my Pi3, 2.15.3 doesn't. Noticed a significant difference in file size. Maybe that's a hint.
> 
> Yes - noticed this too. Is this something specific to the code changes, or perhaps is this a build script / library requirement issue?
> 
> Would it be possible to get a 64-bit Pi version included in the build scripts? Companion Pi requires a Pi4/400 4GB/8GB, and is unsupported on other variants. I'm assuming there's an implicit 64-bit requirement as well (but I could be mistaken).

Hello!

Looping back about this - would it be possible to get it built for aarch64 as well as armv7l?  An aarch64 version for the Pi4 would work on the CompanionPi image.  CompPi+H2R would be a great and easy fit - though I'm not sure how you actually do your builds so I'm not sure if I'm asking for a lot :-/

Unfortunately, all the CompanionPi material is geared towards the 64-bit lite/desktop install, as it needs at least 4gb to run.

**Comment by iamjohnbarker** [comment]
@MarcusAlzona I'll have a think about it and see if I can slot it into the current workflow.

**Comment by mnryan** [comment]
@iamjohnbarker +1 for getting this to run on a Pi4/400 64bit. I would have a very good reason to make a bunch of purchases happen if 64bit support is added.

**Comment by MarcusAlzona** [comment]
> @MarcusAlzona I'll have a think about it and see if I can slot it into the current workflow.

Just curious @iamjohnbarker  - how do you actually get the Pi builds into the workflow?  Is it designating another build target for a multi-platform build tool?  Or did you need to setup a Pi4 with an armv7l OS architecture, and would need a separate Pi4 with an aarch64 running (and/or an SD card swap 😃 ).

**Comment by iamjohnbarker** [comment]
Yes @MarcusAlzona - I got it workong on a pi with armv7l just fine

**Comment by champagj** [comment]
Back at video production after a pause doing "regular-work". My workflow ATEM + CompanionPi + PlayOutBeePi  could really use an H2R running on Pi as well.
Finally got a PI5 which really adds power to make this work nicely. 
It appears 64bit would be the way to go. 
@iamjohnbarker happy to beta-test and pay for a new pro-pi-64bits version!

**Comment by iamjohnbarker** [comment]
@champagj I will try building for `arm64` in the next version (likely mid January) and we can see if it works as expected!

**Comment by ferramarco** [comment]
> > Hi, I have an OrangePi 5, I already installed companion, is it possible to install H2R?
> 
> I've not tried the OrangePi myself, however if you are able to install AppImages then you should be good @ferramarco!

Hi John, I'm back to ask you help.
I need to install on Orangepi, I've tried many times with some os, like Ubuntu etc. but I can't install. It generates a file with some simbols and "(invalid encoding)"  
I've already installed also on a Raspberry pi5, but sometimes it generates some error and companion doesn't work properly, I had to restart it during a Live.
I have a lot of Orangepi5 and I need a way to try to install it on. Now they are running Ubuntu 24.04 and the feeling is that is more stable
Can you help me?
